realsense2_camera/src/base_realsense_node.cpp

```diff
@@ -917,7 +921,10 @@ void BaseRealSenseNode::calcAndPublishStaticTransform(const rs2::stream_profile&
     auto Q = rotationMatrixToQuaternion(ex.rotation);
     Q = quaternion_optical * Q * quaternion_optical.inverse();
 
-    float3 trans{ex.translation[0], ex.translation[1], ex.translation[2]};
+    // Invert x,y,z of extrinsic translation before sending this as TF translation
+    // because extrinsic from A to B is the position of A relative to B
+    // while TF from A to B is the transformation to be done on A to get to B
+    float3 trans{-ex.translation[0], -ex.translation[1], -ex.translation[2]};
```

Comment:
What about the rotation?
I think a more correct way to handle it would be to replace line #906: ex = base_profile.get_extrinsics_to(profile);
with: ex = profile.get_extrinsics_to(base_profile);

In fact, I see it was that way before a PR from september 2022: https://github.com/IntelRealSense/realsense-ros/commit/61406afa1e32b024992606626d29e57258590640

Maybe that PR was a mistake or did it solve a real issue?

Reply:
After discussing this offline, I understood why the change was introduced in the first place. Samer will check if the "get_extrinsics_to()" function call is "heavy" and if not, we might have one for publishing to the extrinsic topic and another one with opposite order of frames to publish to the tf topic.

Reply:
We will have to explain it well in the code, hard to understand why extrinsic msg and static_tf should behave differently.

Reply:
A quick example on D435i why we need the change of #2477:

**Before the fix of PR #2477:**

`administrator@perclnx466 ~/ros2_humble $ ros2 topic echo /camera/extrinsics/depth_to_color
rotation:
- 0.9999583959579468
- -0.008895229548215866
- 0.0020131953060626984
- 0.008895332925021648
- 0.9999604225158691
- -4.254872692399658e-05
- -0.0020127370953559875
- 6.045500049367547e-05
- 0.9999979734420776
translation:
- -0.0148666612803936
- -0.0008839939837343991
- -0.0005615801201201975
`

**After the fix of PR #2477:**

`administrator@perclnx466 ~/ros2_humble $ ros2 topic echo /camera/extrinsics/depth_to_color
rotation:
- 0.9999583959579468
- 0.008895332925021648
- -0.0020127370953559875
- -0.008895229548215866
- 0.9999604225158691
- 6.045500049367547e-05
- 0.0020131953060626984
- -4.254872692399658e-05
- 0.9999979734420776
translation:
- 0.01485931035131216
- 0.0010161789832636714
- 0.0005317096947692335
`


To make sure which result is the right one, all we need is to compare it to `rs-enumare-devices -c `while running the same camera. Somewhere in the output we get this:

```
Extrinsic from "Depth"    To      "Color" :
 Rotation Matrix:
   0.999958        -0.00889523       0.0020132
   0.00889533       0.99996         -4.25487e-05
  -0.00201274       6.0455e-05       0.999998

 Translation Vector: 0.0148593103513122  0.00101617898326367  0.000531709694769233
```


And that makes sense with the new result after the fix.
If we "imagine" that color sensor is (0,0,0), then Depth is on the right of this sensor by 0.01485 m (~1.5 cm), and this is the result that should be published in the relevant topic of the 
**/camera/extrinsics/depth_to_color**

Reply:
Now, the next explanation is to elaborate more why we need in addition to calculate extrinsic in the opposite way as @Gilaadb suggest.

The first extrinsic we calculated above, was used to publish the depth_to_color topic.
As I described that, and as you see the results, it shows where is "Depth" in the eyes of the "Color".

To calculate TF from depth to color, we need the opposite thing, and that is how to go from Depth To Color ?
We can manipulate the Extrinsic result we already calculated above, and do some inversion, to get the right direction for TFs, but as Gilad suggested, it may also be complicated a little bit, especially with the ROTATION matrix, so he suggested to calculated two things:

ex1 = base_profile.get_extrinsics_to(profile);  // will be used for depth_to_color topic
ex2 = profile.get_extrinsics_to(base_profile); // will be used for the  Depth To Color TF information

Reply:
@Gilaadb ,

After a deeper investigation and testing, I would prefer to calculate the extrinsic only once, then publish it for extrinsic topic, and then to manipulate it (invert Matrix and Translation-Vector) for TF.

I tested with ex1 and ex2, and it looks like it add some time to fetch the extrinsic in both ways, as it goes and enumerate on all extrinsic of the camera (twice).
But I learned from that, that also the Rotation Matrix should be flipped (inverted), so I will fix tomorrow and update the code.

so for example, we get today:

```rotation:
      x: 0.004447687417268753
      y: -2.575120197434444e-05
      z: -0.0010064936941489577
      w: 0.9999895691871643
```

but it should be:
```
 rotation:
      x: -0.004447687417268753
      y: 2.575120197434444e-05
      z: 0.0010064936941489577
      w: 0.9999895691871643
```

so, we have to invert X,Y,Z also in the Rotation